program/src/processor.rs

```diff
@@ -1225,6 +1394,58 @@ impl Processor {
         Ok(())
     }
 
+    #[inline(never)]
+    fn close_spot_open_orders(program_id: &Pubkey, accounts: &[AccountInfo]) -> MangoResult<()> {
+        const NUM_FIXED: usize = 7;
+        let accounts = array_ref![accounts, 0, NUM_FIXED];
+
+        let [
+            mango_group_ai,     // read
+            mango_account_ai,   // write
+            owner_ai,           // write, signer
+            dex_prog_ai,        // read
+            open_orders_ai,     // write
+            spot_market_ai,     // read
+            signer_ai,          // read
+        ] = accounts;
+
+        check!(owner_ai.is_signer, MangoErrorCode::InvalidSignerKey)?;
+
+        let mango_group = MangoGroup::load_checked(mango_group_ai, program_id)?;
+        check_eq!(dex_prog_ai.key, &mango_group.dex_program_id, MangoErrorCode::InvalidProgramId)?;
+        check_eq!(signer_ai.key, &mango_group.signer_key, MangoErrorCode::InvalidParam)?;
+
+        let market_index = mango_group
+            .find_spot_market_index(spot_market_ai.key)
+            .ok_or(throw_err!(MangoErrorCode::InvalidMarket))?;
+
+        let mut mango_account =
+            MangoAccount::load_mut_checked(mango_account_ai, program_id, &mango_group_ai.key)?;
+        check_eq!(&mango_account.owner, owner_ai.key, MangoErrorCode::InvalidOwner)?;
+        check!(!mango_account.being_liquidated, MangoErrorCode::BeingLiquidated)?;
+        check!(!mango_account.is_bankrupt, MangoErrorCode::Bankrupt)?;
+        check_eq!(
+            &mango_account.spot_open_orders[market_index],
+            open_orders_ai.key,
+            MangoErrorCode::InvalidOwner
+        )?;
+
+        let signers_seeds = gen_signer_seeds(&mango_group.signer_nonce, mango_group_ai.key);
+        invoke_close_open_orders(
+            dex_prog_ai,
+            open_orders_ai,
+            signer_ai,
+            owner_ai,
+            spot_market_ai,
+            &[&signers_seeds],
+        )?;
+
+        mango_account.spot_open_orders[market_index] = Pubkey::default();
+        mango_account.in_margin_basket[market_index] = false;
```

Comment:
in_margin_basket should only be modified in state.rs -- I could make it a private var, but easier to just follow this rule for now. A bug that could happen here is if an OpenOrders is in_margin_basket, but it can actually be closed (we know this can happen with current protocol). Then invoke_close_open_orders will work and in_margin_basket will be set to false, but num_in_margin_basket will not equal sum(in_margin_basket) which is required.